# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](tttboard.jpg)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [315]:
# your code here
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
data = pd.read_csv('tic-tac-toe.csv')

print(data.head())
data.dtypes

  TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True


TL       object
TM       object
TR       object
ML       object
MM       object
MR       object
BL       object
BM       object
BR       object
class      bool
dtype: object

In [316]:
#Convert from categorical to nuerical using EnconderLabel
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
data = data.apply(le.fit_transform) 
print(data.head())
data.dtypes


   TL  TM  TR  ML  MM  MR  BL  BM  BR  class
0   2   2   2   2   1   1   2   1   1      1
1   2   2   2   2   1   1   1   2   1      1
2   2   2   2   2   1   1   1   1   2      1
3   2   2   2   2   1   1   1   0   0      1
4   2   2   2   2   1   1   0   1   0      1


TL       int32
TM       int32
TR       int32
ML       int32
MM       int32
MR       int32
BL       int32
BM       int32
BR       int32
class    int64
dtype: object

In [317]:
data.isnull().sum()

TL       0
TM       0
TR       0
ML       0
MM       0
MR       0
BL       0
BM       0
BR       0
class    0
dtype: int64

In [318]:
#Separate Features and target
features = data.drop('class', axis=1)
target = data['class']

n_cols = features.shape[1] # number of predictors
print(n_cols)

9


## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [319]:
# your code here
#Split the data into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.3, random_state=42)


In [320]:
print("Training and testing data shapes:", X_train.shape,  y_train.shape, y_test.shape,X_test.shape,)#add a string to the print statement)


Training and testing data shapes: (670, 9) (670,) (288,) (288, 9)


In [321]:
#Creat Model
from keras.models import Sequential
from keras.layers import Dense , Dropout
from keras.optimizers import Adam


# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(64, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(16, activation='relu'))
    model.add(Dense(n_cols, activation='softmax'))
    
    # compile model
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model


## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [322]:
# your code here
model = regression_model()
model.fit(X_train, y_train, epochs=100, verbose=1)

#Make predition
y_pred = model.predict(X_test)

# Evaluate the model
score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

# Save the model
model.save('tic-tac-toe.model')

Epoch 1/100
21/21 [==============================] - 1s 6ms/step - loss: 1.9767 - accuracy: 0.2612
Epoch 2/100
21/21 [==============================] - 0s 7ms/step - loss: 1.4039 - accuracy: 0.5806
Epoch 3/100
21/21 [==============================] - 0s 6ms/step - loss: 0.8891 - accuracy: 0.6030
Epoch 4/100
21/21 [==============================] - 0s 5ms/step - loss: 0.6938 - accuracy: 0.6179
Epoch 5/100
21/21 [==============================] - 0s 4ms/step - loss: 0.6413 - accuracy: 0.6522
Epoch 6/100
21/21 [==============================] - 0s 5ms/step - loss: 0.6217 - accuracy: 0.6761
Epoch 7/100
21/21 [==============================] - 0s 4ms/step - loss: 0.6096 - accuracy: 0.6716
Epoch 8/100
21/21 [==============================] - 0s 4ms/step - loss: 0.6006 - accuracy: 0.6970
Epoch 9/100
21/21 [==============================] - 0s 5ms/step - loss: 0.5943 - accuracy: 0.6985
Epoch 10/100
21/21 [==============================] - 0s 5ms/step - loss: 0.5807 - accuracy: 0.7104
Epoch 11/

In [323]:
score = model.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Test loss: 0.3617425560951233
Test accuracy: 0.8298611044883728


## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [324]:
# your code here
#Creat Model
import tensorflow 
# define regression model
def regression_model2():
    # create model
    model2 = Sequential()
    model2.add(Dense(64, activation='relu', input_shape=(n_cols,)))
    model2.add(Dropout(0.2))
    model2.add(Dense(32, activation='relu'))
    model2.add(Dropout(0.2))
    model2.add(Dense(16, activation='relu'))
    model2.add(Dropout(0.2))
    model2.add(Dense(n_cols, activation='softmax'))
    
    # compile model
    #my_opt = 'adam'
    my_opt = tensorflow.keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=0.01)
    #my_opt = Adam(learning_rate=0.001, beta_1=0.9, beta_2=0.999, epsilon=0.01)
    model2.compile(optimizer=my_opt, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    return model2

In [325]:
model2 = regression_model2()
model2.fit(X_train, y_train, epochs=1000, verbose=2)

#Make predition
y_pred2 = model2.predict(X_test)



Epoch 1/1000


c:\Users\edwar\anaconda3\envs\py_gpu\lib\site-packages\keras\optimizers\optimizer_v2\adam.py:110: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(Adam, self).__init__(name, **kwargs)


21/21 - 1s - loss: 2.3892 - accuracy: 0.0478 - 580ms/epoch - 28ms/step
Epoch 2/1000
21/21 - 0s - loss: 2.1787 - accuracy: 0.1687 - 124ms/epoch - 6ms/step
Epoch 3/1000
21/21 - 0s - loss: 1.9739 - accuracy: 0.2612 - 128ms/epoch - 6ms/step
Epoch 4/1000
21/21 - 0s - loss: 1.6952 - accuracy: 0.3239 - 109ms/epoch - 5ms/step
Epoch 5/1000
21/21 - 0s - loss: 1.4794 - accuracy: 0.3478 - 100ms/epoch - 5ms/step
Epoch 6/1000
21/21 - 0s - loss: 1.2058 - accuracy: 0.4119 - 97ms/epoch - 5ms/step
Epoch 7/1000
21/21 - 0s - loss: 1.0721 - accuracy: 0.4896 - 103ms/epoch - 5ms/step
Epoch 8/1000
21/21 - 0s - loss: 1.0158 - accuracy: 0.5642 - 103ms/epoch - 5ms/step
Epoch 9/1000
21/21 - 0s - loss: 0.9510 - accuracy: 0.5687 - 100ms/epoch - 5ms/step
Epoch 10/1000
21/21 - 0s - loss: 0.9169 - accuracy: 0.5731 - 106ms/epoch - 5ms/step
Epoch 11/1000
21/21 - 0s - loss: 0.8832 - accuracy: 0.5836 - 98ms/epoch - 5ms/step
Epoch 12/1000
21/21 - 0s - loss: 0.8795 - accuracy: 0.5642 - 94ms/epoch - 4ms/step
Epoch 13/1000
21

In [328]:
# Evaluate the model
score2 = model2.evaluate(X_test, y_test, verbose=0)
print('Test loss:', score2[0])
print('Test accuracy:', score2[1])

# Save the model
#model2.save('tic-tac-toe.model2')

Test loss: 0.15634438395500183
Test accuracy: 0.9618055820465088


In [327]:
model2.save('tic-tac-toe.model2')

INFO:tensorflow:Assets written to: tic-tac-toe.model2\assets


**Which approach(es) did you find helpful to improve your model performance?**

# your answer here
To improve the model, I added dropout layers, change the opimizer and added one more hidden layer with 32 neurons. The dropiut layers are used to prevent overfitting.

We also change the optimizer to Adam with a learning rate of 0.001, beta_1=0.9, beta_2=0.999, epsilon=0.01. 
This optimizer is more efficient than the default one. This change significanlty changed the model performance.
The accuracy increased to 0.96 and the loss decreased to 0.1.
